### PR TITLE
Add support for Stripe Connect. #1677

### PR DESCRIPTION
**Pull request by mindctrl**
#1677

### CODE REVIEW

File: includes/admin/settings/settings.php [hunk 1]
```diff
@@ -281,6 +281,43 @@ function rcp_settings_page() {
...
+
+						if( rcp_is_sandbox() ) {
+							$rcp_options['stripe_test_publishable'] = $data['publishable_key'];
+							$rcp_options['stripe_test_secret'] = $data['secret_key'];
+						} else {
+							$rcp_options['stripe_live_publishable'] = $data['publishable_key'];
+							$rcp_options['stripe_live_secret'] = $data['secret_key'];
+						}
+						update_option( 'rcp_settings', $rcp_options );
+						wp_redirect( esc_url( admin_url( 'admin.php?page=rcp-settings#payments' ) ) );
```
Comment (nosegraze) [inline includes/admin/settings/settings.php]: I think you're supposed to use `esc_url_raw()` for redirects.

File: includes/admin/settings/settings.php [hunk 2]
```diff
@@ -281,6 +281,43 @@ function rcp_settings_page() {
...
+						$data = json_decode( $response['body'], true )['data'];
+
+						global $rcp_options;
+
+						if( rcp_is_sandbox() ) {
+							$rcp_options['stripe_test_publishable'] = $data['publishable_key'];
+							$rcp_options['stripe_test_secret'] = $data['secret_key'];
+						} else {
+							$rcp_options['stripe_live_publishable'] = $data['publishable_key'];
+							$rcp_options['stripe_live_secret'] = $data['secret_key'];
```
Comment (nosegraze) [inline includes/admin/settings/settings.php]: We should use `sanitize_text_field()` on all these values.

File: includes/admin/settings/settings.php [hunk 3]
```diff
@@ -348,13 +385,26 @@ function rcp_settings_page() {
...
+							<td>
+							<?php
+							if( ( empty( $rcp_options['stripe_test_publishable'] ) && rcp_is_sandbox() ) || ( empty( $rcp_options['stripe_live_publishable'] ) && ! rcp_is_sandbox() ) ):
+								$stripe_connect_url = add_query_arg( array(
+									'live_mode' => (int) ! rcp_is_sandbox(),
+									'state' => str_pad( wp_rand( wp_rand(), PHP_INT_MAX ), 100, wp_rand(), STR_PAD_BOTH ),
+									'customer_site_url' => admin_url( 'admin.php?page=rcp-settings' ),
+								), 'https://restrictcontentpro.com/?rcp_gateway_connect_init=stripe_connect' );
+								?>
+								<a href="<?php echo esc_url_raw( $stripe_connect_url ); ?>" class="rcp-stripe-connect"><span><?php _e( 'Connect with Stripe', 'rcp' ); ?></span></a>
```
Comment (nosegraze) [inline includes/admin/settings/settings.php]: This should be `esc_url()` - not `esc_url_raw()`.

Comment (mindctrl) [reply includes/admin/settings/settings.php]: `esc_url()` converts the & entities and Pagely was dying on that, though right now it doesn't seem to be.   ¯\_(ツ)_/¯ 

I've been using multiple browsers and don't remember what was going on at this point. I'll do some more testing.

**Comment by mindctrl** [comment]
Fixed the merge conflict, but I'm going to wait to merge this. I need to get some clarity from Stripe on something first. I've emailed them. I'll give an update here when I hear back.

**Comment by mindctrl** [comment]
This needs to be updated to also pass the customer's Stripe customer ID during charge/subscription creation. I'll get this done by Monday.

**Comment by mindctrl** [comment]
@nosegraze can you review and test the latest changes?

**Comment by mindctrl** [comment]
You'll need to reconnect your Stripe account.

**Comment by nosegraze** [comment]
I'm now getting the following error right below the `Completing your Stripe account setup...` message:

```
[11-Apr-2018 09:45:46 UTC] PHP Warning:  Cannot modify header information - headers already sent by (output started at C:\wamp\www\rcp\wp-includes\class.wp-styles.php:225) in C:\wamp\www\rcp\wp-includes\pluggable.php on line 1216

[11-Apr-2018 09:45:46 UTC] PHP Stack trace:

[11-Apr-2018 09:45:46 UTC] PHP   1. {main}() C:\wamp\www\rcp\wp-admin\admin.php:0

[11-Apr-2018 09:45:46 UTC] PHP   2. do_action() C:\wamp\www\rcp\wp-admin\admin.php:224

[11-Apr-2018 09:45:46 UTC] PHP   3. WP_Hook->do_action() C:\wamp\www\rcp\wp-includes\plugin.php:453

[11-Apr-2018 09:45:46 UTC] PHP   4. WP_Hook->apply_filters() C:\wamp\www\rcp\wp-includes\class-wp-hook.php:310

[11-Apr-2018 09:45:46 UTC] PHP   5. call_user_func_array:{C:\wamp\www\rcp\wp-includes\class-wp-hook.php:286}() C:\wamp\www\rcp\wp-includes\class-wp-hook.php:286

[11-Apr-2018 09:45:46 UTC] PHP   6. rcp_settings_page() C:\wamp\www\rcp\wp-includes\class-wp-hook.php:286

[11-Apr-2018 09:45:46 UTC] PHP   7. wp_redirect() C:\wamp\www\rcp\wp-content\plugins\restrict-content-pro\includes\admin\settings\settings.php:320

[11-Apr-2018 09:45:46 UTC] PHP   8. header() C:\wamp\www\rcp\wp-includes\pluggable.php:1216
```

**Comment by nosegraze** [comment]
I guess due to `wp_redirect()` being inside the `<body>` tag and after the headers have been set.

**Comment by mindctrl** [comment]
Pippin saw this too. What's weird is it has not happened once on my site. I'll rework it.

**Comment by nosegraze** [comment]
Yeah it's weird because I didn't get it last time I tested.

**Comment by mindctrl** [comment]
@nosegraze I've updated this PR and the server side code and have tested it locally. Can you do another round of tests?

**Comment by nosegraze** [comment]
I tested a cancel and a connection - both worked well!

Just a random thought... it doesn't have to be done now, but maybe in the future, could we use this process to set up webhooks for people automatically?

**Comment by nosegraze** [comment]
Shucks, that'd be so useful.

**Comment by mindctrl** [comment]
Emailed Stripe again to get a final confirmation that things look right on their end and that we don't need to do anything else. Once that's confirmed, we can merge this.

**Pull request merged**